### PR TITLE
[bugfix]: add pre-check that If the cluster already exists

### LINKED ISSUE #2022: [sealer run]check if there is a running cluster, if exist, should not mount new image.
### What happen?

对于一个已经部署了集群镜像A的集群1，如果再次运行集群镜像B，那么sealer目前不会检查，然后直接mount了集群镜像B，可能就将镜像A的文件冲掉了。

For a cluster 1 that has already deployed cluster-image A, if cluster-image B is run again, the sealer will  directly mounts cluster image B without checking, and then may overwrite the file of image A.

### Relevant log output?

_No response_

### What you expected to happen?

我希望sealer run时，会提前检查是否有已经部署的集群镜像，如果有，且版本不一致，那么报错，如果版本一致，那么跳过。如果没有，正常安装

I hope that when the sealer runs, it will check whether there is a deployed cluster image. If there is, and the version is not equal, then throw an error. If the version is equal, then skip. If not, install normally

### How to reproduce it (as minimally and precisely as possible)?

_No response_

### Anything else we need to know?

_No response_

### What is the version of Sealer you using?

_No response_

### What is your OS environment?

_No response_

### What is the Kernel version?

_No response_

### Other environment you want to tell us?

- Cloud provider or hardware configuration:
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by Stevent-fei**
### Describe what this PR does / why we need it

When the sealer runs, it will check in advance whether there is a deployed cluster image. If there is, and the version is inconsistent, then an error will be reported. If the version is consistent, then skip. If not, install normally

exist：
image

not exist：
image

### Does this pull request fix one issue?

Fixes #2022

### CODE REVIEW

File: cmd/sealer/cmd/cluster/run.go [hunk 1]
```diff
@@ -115,6 +118,24 @@ func NewRunCmd() *cobra.Command {
...
+				file, err := os.Open(common.GetDefaultClusterfile())
+				if err != nil {
+					return err
+				}
+				context, err := ioutil.ReadAll(file)
+				if err != nil {
+					return err
+				}
+
+				if ok := strings.Contains(string(context), args[0]); !ok {
```
Comment (VinceCui) [inline cmd/sealer/cmd/cluster/run.go:131]: No need check if cluster image same, if not same, the user also can't run cluster.

Comment (VinceCui) [reply cmd/sealer/cmd/cluster/run.go:131]: Uer API to marshal it.

**Comment by kakaZhou719** [comment]
@Stevent-fei i have two questions:

1. how to determine there is an existed cluster?
2. should we add this pre-check logic to all create process?

File: cmd/sealer/cmd/cluster/run.go [hunk 2]
```diff
@@ -115,6 +117,30 @@ func NewRunCmd() *cobra.Command {
 					app, extension, nil, imageEngine, runFlags.Mode)
 			}
 
+			if exist := osi.IsFileExist(common.GetDefaultClusterfile()); exist {
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/cluster/run.go:120]: not use file existing  to do this pre-check, try get  it from `clusterfile.NewClusterFile(nil)`,
 if not found, no cluster.
 if have ,determine whether the image name is the same one.

**Review by Starnop — APPROVED** [review]
LGTM

File: cmd/sealer/cmd/cluster/run.go [hunk 3]
```diff
@@ -427,6 +433,19 @@ func installApplication(appImageName string, envs []string, app *v2.Application,
 	return nil
 }
 
+func checkClusterIsExists(cf clusterfile.Interface) error {
+	rawCluster := cf.GetCluster()
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/run.go:437]: rawCluster/desiredCluster/g

File: cmd/sealer/cmd/cluster/run.go [hunk 4]
```diff
@@ -427,6 +433,19 @@ func installApplication(appImageName string, envs []string, app *v2.Application,
...
+	file, err := clusterfile.NewClusterFile(nil)
+	if err == nil {
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/run.go:439]: if err!=nil && match some regular，return nil

File: cmd/sealer/cmd/cluster/run.go [hunk 5]
```diff
@@ -510,3 +517,16 @@ func prepareMaterials(infraDriver infradriver.InfraDriver, imageEngine imageengi
 	}
 	return nil
 }
+
+func checkClusterIsExists(cf clusterfile.Interface) error {
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/cluster/run.go:521]: not good use err to identify the check result. i assume  we could return passed result and error.

**Comment by kakaZhou719** [comment]
@Stevent-fei . it ok for me to do pre-check, such as cluster image . however , that is just one item need to check . it is better that if we add a check list to do this.

File: cmd/sealer/cmd/cluster/run.go [hunk 6]
```diff
@@ -510,3 +518,16 @@ func prepareMaterials(infraDriver infradriver.InfraDriver, imageEngine imageengi
 	}
 	return nil
 }
+
+func checkClusterIsExists(cf clusterfile.Interface) (string, error) {
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/cluster/run.go:522]: func checkClusterIsExists(cluster v2.cluster) (string, error) {

Comment (kakaZhou719) [inline cmd/sealer/cmd/cluster/run.go:522]: move this func to different file Or utils, and add UT test.

File: cmd/sealer/cmd/cluster/run.go [hunk 7]
```diff
@@ -510,3 +518,16 @@ func prepareMaterials(infraDriver infradriver.InfraDriver, imageEngine imageengi
...
+	rawCluster := cf.GetCluster()
+	clusterFile, _, err := clusterfile.GetActualClusterFile()
+	if err == nil {
+		cluster := clusterFile.GetCluster()
+		if cluster.Spec.Image != rawCluster.Spec.Image {
+			return "exist", fmt.Errorf("the cluster image already exists, please uninstall the current cluster and install another version of the cluster image")
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/cluster/run.go:528]: "exist" is not good, you should use const such as "SameImageName" Or something to identity it

File: cmd/sealer/cmd/cluster/run.go [hunk 8]
```diff
@@ -235,6 +237,12 @@ func runWithClusterfile(clusterFile string, runFlags *types.RunFlags) error {
 }
 
 func runClusterImage(imageEngine imageengine.Interface, cf clusterfile.Interface, imageSpec *imagev1.ImageSpec, mode string) error {
+	//check cluster
+	checkClusterFile, err := checkClusterIsExists(cf)
+	if checkClusterFile == "exist" && err != nil {
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/cluster/run.go:242]: why &&? Return err if it is not nil.

File: cmd/sealer/cmd/utils/cluster.go [hunk 9]
```diff
@@ -256,3 +259,15 @@ func GetClusterClient() *k8s.Client {
 	}
 	return nil
 }
+
+func CheckClusterIsExists(cluster v2.Cluster) (string, error) {
+	clusterFile, _, err := clusterfile.GetActualClusterFile()
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/utils/cluster.go:264]: should retrun err fast

File: cmd/sealer/cmd/utils/cluster.go [hunk 10]
```diff
@@ -256,3 +259,15 @@ func GetClusterClient() *k8s.Client {
...
+	if err == nil {
+		newClusterfile := clusterFile.GetCluster()
+		if cluster.Spec.Image != newClusterfile.Spec.Image {
+			return clusterExists, fmt.Errorf("the cluster image already exists, please uninstall the current cluster and install another version of the cluster image")
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/utils/cluster.go:268]: should not retrun err

**Pull request closed without merging**